<a href="https://colab.research.google.com/github/USMS-FPK-IA-GI-2025/controle-01-numpy-and-matplotlib-iris-Laylaouyoussef68/blob/main/notebooks/controle_numpy_matplotlib.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Contrôle 01 — Data Science avec NumPy et Matplotlib
**Master 1 — Intelligence Artificielle & Génie Informatique (FPK / USMS)**

Nom & Prénom :  **OUYOUSSEF** **LAYLA**

Filière :  **MASTER** **SIIA**

Année universitaire : **2025**-**2026**

---

**Règles :**
- Bibliothèques autorisées : NumPy, Matplotlib
- Toute autre bibliothèque est interdite
- Chaque question doit être traitée dans une cellule dédiée
- Chaque résultat doit être brièvement interprété


# Partie A — Importation du Dataset, Arrays, Premières Manipulations

### A1. Importation du dataset
Importez NumPy et Matplotlib.  
Chargez le fichier `iris.csv` avec `np.genfromtxt` (`delimiter=','`, `skip_header=1`).  
Stockez les données dans un array nommé `data` et affichez sa `shape`.


In [18]:
#importation du Numpy et Matplotlib
import numpy as np
import matplotlib as plt

In [19]:
# chargez le fichier csv
data = np.genfromtxt('/content/Iris.csv',delimiter=',',skip_header=1)
print(data.shape)

(150, 6)


L'output **(150, 6)** signifie que l'array data contient **150 lignes** (soit les 150 échantillons de fleurs d'Iris) et **6 colonnes** (les 4 mesures de sépale/pétale, la colonne de l'espèce variety, plus une colonne d'index/ID)

### A2. Séparation des features et des espèces
Séparez :
- les 4 premières colonnes numériques dans un array `features`
- la dernière colonne (espèces) dans un array `species`
Utilisez le slicing.


In [33]:
# Separez les 4 premieres colonnes dans features
features = data[:,1:5]
#array species contiet la derniere colonne
species = data[:,-1]
#afficher taille de features
print("shape de features : ",features.shape)
print("\n")
#afficher taille de species
print("shape de species : ",species.shape)

shape de features :  (150, 4)


shape de species :  (150,)


* Le résultat de **features.shape** est **(150, 4)**, ce qui confirme que l'opération de slicing a correctement extrait les 4 colonnes

* L'affichage de ***species.shape*** est ***(150,)***, ce qui indique qu'un vecteur unidimensionnel de 150 éléments a été créé en extrayant la dernière colonne.


### A3. Indexing
Affichez les 5 premières lignes de `features` en utilisant l’indexing.


In [34]:
# l'affichage des 5 premieres lignes de features
print(features[0:5,:])


[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


Chacune des 5 lignes affichées représente un échantillon de fleur d'Iris, et les **quatre** **valeurs** sur chaque ligne correspondent aux caractéristiques numériques suivantes :


*  sepal_length
*  sepal_width
*  petal_length
*  petal_width

Cet affichage confirme que le slicing effectué à l'étape A2 a été correct, et que les données numériques sont correctement structurées dans l'array features


### A4. Extraction d’une colonne
Créez un array `sepal_length` contenant uniquement la première colonne de `features`.  
Affichez sa `size` et son `dtype`.


In [35]:
# creation d'array sepal_length
sepal_length = features[:,0:1]
#affichage des 5 premiere ligne de sepal_length
print(sepal_length[0:5,])
#afficher size de sepal_length
print("size de sepal_length : ",sepal_length.size)
print("type de sepal_length : ",sepal_length.dtype)

[[5.1]
 [4.9]
 [4.7]
 [4.6]
 [5. ]]
size de sepal_length :  150
type de sepal_length :  float64


L'affichage de sepal_length.size est **150**, ce qui confirme que nous avons correctement extrait la colonne entière des longueurs de sépales pour les 150 observations. Le type de données (**dtype**) **float64** indique que les valeurs sont stockées comme des nombres à virgule flottante

### A5. Slicing par espèce
Utilisez le slicing pour extraire les 50 premières lignes correspondant à l’espèce *Setosa*.  
Stockez-les dans `setosa_features`.


In [36]:
#extraire les 50 premieres lignes correspondant a Setosa
setosa_features = features[0:50,]
print("la taille de setosa_features :", setosa_features.shape)

la taille de setosa_features : (50, 4)


Le résultat **setosa_features**.**shape** est (50, 4). Cette forme confirme que nous avons correctement isolé les 50 premières observations (les 50 fleurs de l'espèce Setosa) tout en conservant leurs **4 caractéristiques** numériques

# Partie B — Statistiques Basiques et Opérations Mathématiques

### B6. Moyenne par colonne
Calculez la moyenne de chaque colonne de `features` avec `np.mean(axis=0)`.


In [37]:
#la moyenne de chaque colonne de features
moy =  np.mean(features,axis=0)
print(moy)

[5.84333333 3.054      3.75866667 1.19866667]



Ce vecteur de 4 valeurs représente les caractéristiques moyennes de l'ensemble du dataset Iris :


*   La première valeur (5.84333333 cm) est la **longueur moyenne du sépale**
*   La deuxième valeur (3.054   cm) est **la largeur moyenne du sépale**


*  La troisième valeur ( 3.75866667 cm) est la **longueur moyenne du pétale.**
*  La quatrième valeur (1.19866667 cm) est la **largeur moyenne du pétale.**





### B7. Min et Max
Trouvez la valeur minimale et maximale de la longueur des pétales.


In [40]:
#afficher la valeur minimale de la longueur des petales
min_pet = np.min(features[:,2:3])
print("la valeur minimale de la longueur des petales : ",min_pet)
#afficher la valeur maximale de la longueur des petales
max_pet = np.max(features[:,2:3])
print("la valeur maximale de la longueur des petales est : ",max_pet)

la valeur minimale de la longueur des petales :  1.0
la valeur maximale de la longueur des petales est :  6.9


La **valeur minimale** de **1.0** et la **valeur maximale** de **6.9** centimètres définissent l'intervalle dans lequel se situent toutes les observations de longueurs de pétales

### B9. Normalisation
Normalisez les longueurs des pétales en les divisant par leur maximum.
Stockez le résultat dans `normalized_petals`.


In [43]:
#normalisation
normalized_petals = features[:,2:3]/max_pet
print("longuer des petales normaliser est : \n",normalized_petals)

longuer des petales normaliser est : 
 [[0.20289855]
 [0.20289855]
 [0.1884058 ]
 [0.2173913 ]
 [0.20289855]
 [0.24637681]
 [0.20289855]
 [0.2173913 ]
 [0.20289855]
 [0.2173913 ]
 [0.2173913 ]
 [0.23188406]
 [0.20289855]
 [0.15942029]
 [0.17391304]
 [0.2173913 ]
 [0.1884058 ]
 [0.20289855]
 [0.24637681]
 [0.2173913 ]
 [0.24637681]
 [0.2173913 ]
 [0.14492754]
 [0.24637681]
 [0.27536232]
 [0.23188406]
 [0.23188406]
 [0.2173913 ]
 [0.20289855]
 [0.23188406]
 [0.23188406]
 [0.2173913 ]
 [0.2173913 ]
 [0.20289855]
 [0.2173913 ]
 [0.17391304]
 [0.1884058 ]
 [0.2173913 ]
 [0.1884058 ]
 [0.2173913 ]
 [0.1884058 ]
 [0.1884058 ]
 [0.1884058 ]
 [0.23188406]
 [0.27536232]
 [0.20289855]
 [0.23188406]
 [0.20289855]
 [0.2173913 ]
 [0.20289855]
 [0.68115942]
 [0.65217391]
 [0.71014493]
 [0.57971014]
 [0.66666667]
 [0.65217391]
 [0.68115942]
 [0.47826087]
 [0.66666667]
 [0.56521739]
 [0.50724638]
 [0.60869565]
 [0.57971014]
 [0.68115942]
 [0.52173913]
 [0.63768116]
 [0.65217391]
 [0.5942029 ]
 [0.65217

L'opération a consisté à normaliser les longueurs de pétales en divisant chaque valeur par la valeur maximale trouvée précédemment (max_pet = 6.9). Le résultat stocké dans normalized_petals est un nouvel array où toutes les valeurs sont désormais comprises entre 0 et 1

### B10. Somme cumulée
Calculez la somme cumulée des largeurs des pétales pour Versicolor.
Affichez les 10 premières valeurs.


In [68]:
#la somme cummulee
#stocker les largeurs des petales dans ver_pet_wi
ver_pet_wi = features[50:100,3:4]
print("size de ver_pet_wi : ",ver_pet_wi.shape)
sum_cumm = np.cumsum(ver_pet_wi)
print("la somme cumuee est : ",sum_cumm)

size de ver_pet_wi :  (50, 1)
la somme cumuee est :  [ 1.4  2.9  4.4  5.7  7.2  8.5 10.1 11.1 12.4 13.8 14.8 16.3 17.3 18.7
 20.  21.4 22.9 23.9 25.4 26.5 28.3 29.6 31.1 32.3 33.6 35.  36.4 38.1
 39.6 40.6 41.7 42.7 43.9 45.5 47.  48.6 50.1 51.4 52.7 54.  55.2 56.6
 57.8 58.8 60.1 61.3 62.6 63.9 65.  66.3]


### B10. Somme cumulée
Calculez la somme cumulée des largeurs des pétales pour Versicolor.
Affichez les 10 premières valeurs.


In [69]:
#la somme cummulee
#stocker les largeurs des petales dans ver_pet_wi
ver_pet_wi = features[50:100,3:4]
print("size de ver_pet_wi : ",ver_pet_wi.shape)
sum_cumm = np.cumsum(ver_pet_wi)
print("la somme cumuee est : ",sum_cumm)

size de ver_pet_wi :  (50, 1)
la somme cumuee est :  [ 1.4  2.9  4.4  5.7  7.2  8.5 10.1 11.1 12.4 13.8 14.8 16.3 17.3 18.7
 20.  21.4 22.9 23.9 25.4 26.5 28.3 29.6 31.1 32.3 33.6 35.  36.4 38.1
 39.6 40.6 41.7 42.7 43.9 45.5 47.  48.6 50.1 51.4 52.7 54.  55.2 56.6
 57.8 58.8 60.1 61.3 62.6 63.9 65.  66.3]


L'opération np.cumsum() sur la largeur des pétales de l'espèce Versicolor (lignes 50 à 99) a généré un vecteur de 50 éléments où chaque valeur représente la somme progressive des largeurs de pétales observées jusqu'à cet échantillon. Le résultat final, 66.3, est la somme totale des largeurs de pétales pour les 50 fleurs Versicolor.

# Partie C — Transformations Avancées avec NumPy

### C11. Espèces uniques
Listez les espèces uniques avec `np.unique` et comptez leurs occurrences.


In [ ]:
# Votre code ici


### C12. Reshape par espèce
Transformez `features` en un array de forme `(3, 50, 4)` nommé `features_by_species`.


In [ ]:
# Votre code ici


### C13. Concaténation
Concaténez horizontalement les moyennes par colonne avec un array de zéros.


In [ ]:
# Votre code ici


### C14. Data cleaning (simulation)
Simulez des valeurs NaN et remplacez-les avec `np.nan_to_num`.


In [ ]:
# Votre code ici


### C15. Broadcasting
Ajoutez 1 à toutes les valeurs de la première colonne de `features`.


In [ ]:
# Votre code ici


### C16. np.where
Créez un array binaire indiquant si la longueur des pétales est supérieure à 4.0.


In [ ]:
# Votre code ici


# Partie D — Visualisation avec Matplotlib (Plots Simples)

### D17. Line plot
Line plot des longueurs des sépales pour les 50 premières fleurs.


In [ ]:
# Votre code ici


### D18. Scatter plot
Scatter plot : longueur vs largeur des pétales.


In [ ]:
# Votre code ici


### D19. Bar chart
Bar chart des moyennes des quatre caractéristiques.


In [ ]:
# Votre code ici


### D20. Histogram
Histogramme de la largeur des sépales (10 bins).


In [ ]:
# Votre code ici


### D21. Boxplot par espèce
Boxplot des longueurs des pétales par espèce.


In [ ]:
# Votre code ici


# Partie E — Visualisations Avancées

### E22. Subplots comparatifs
Deux scatter plots côte à côte : Setosa vs Versicolor.


In [ ]:
# Votre code ici


### E23. Grille 2x2
Histogrammes des 4 caractéristiques avec `plt.subplots(2,2)`.


In [ ]:
# Votre code ici


### E24. Bar chart avec error bars
Moyennes par espèce avec écart-type.


In [ ]:
# Votre code ici


### E25. Scatter coloré par espèce
Scatter plot avec couleurs par espèce.


In [ ]:
# Votre code ici


### E26. Line plot superposé
Cumsum des largeurs des pétales par espèce.


In [ ]:
# Votre code ici


# Partie F — Mini-Analyse Finale

### F27. Corrélations
Calculez la matrice de corrélation avec `np.corrcoef`.


In [ ]:
# Votre code ici


### F28. Visualisation des corrélations
Visualisez les corrélations sous forme de bar chart ou scatter.


In [ ]:
# Votre code ici


### F29. Variance par espèce
Identifiez l’espèce avec la plus grande variance globale et visualisez ses distributions.


In [ ]:
# Votre code ici


### F30. Conclusion
Rédigez une conclusion contenant **3 insights** issus de votre analyse.


1) ...
2) ...
3) ...